internal/pkg/composable/controller.go

```diff
@@ -173,7 +173,7 @@ func (c *controller) Run(ctx context.Context) error {
 					}
 				}()
 
-				close(c.ch)
+				//close(c.ch)
```

Comment:
Is this intentionally commented out?

Reply:
Ah, forgot to mention that in the PR description, I ran into a lot of `panic on closed channel` errors after I fixed the initial deadlock. Unless we get a Context Cancel error, agent will try to restart, and that restart will (at least in this case) cause that `close()` to be called twice. I'm not sure if the close() is needed, but removing it at this point isn't strictly needed, as my fix was to make sure that an error in one of the subcomponents doesn't trigger a restart. However, that issue can probably crop up somewhere else...

Reply:
@fearful-symmetry  did you manage to see if we got the double close because we are recycling/reusing a closed channel or if it's some sort of race condition ? 
If we wanna be safe we could put a recover() around the close and log maybe part of the stack that brought us here?

Reply:
If the scenario in https://github.com/elastic/elastic-agent/pull/2352#issuecomment-1460666211 is accurate, it means that we try to reuse a closed channel: this will trigger a panic also with a simple write

This comment would avoid the panic during the shutdown/cleanup but the real issue is that we are trying to reuse a controller that already failed once without restoring its initial working state (which looks to me the real underlying issue)

In the end if no reader is blocking on reading from `c.ch` at this stage, commenting out the close() should have no adverse effects but I think we would be safer with using a `sync.Once` something like:

```suggestion
				c.once.Do(func(){close(c.ch)})
```

where the `Once` instance and the channel `ch` share the same lifetime

This however is non-blocking for this PR

Reply:
That is https://github.com/elastic/elastic-agent/pull/2349 for reference